circleci: failed tests: TestReplicaDataIterator

The following test appears to have failed:

[#21078](https://circleci.com/gh/cockroachdb/cockroach/21078):

```
--- PASS: TestReplicaDataIteratorEmptyRange (0.02s)
=== RUN   TestReplicaDataIterator
W160808 20:17:22.074712 gossip/gossip.go:1015  not connected to cluster; use --join to specify a connected node
I160808 20:17:22.075031 storage/engine/rocksdb.go:353  opening in memory rocksdb instance
I160808 20:17:22.078380 storage/store.go:2718  system config not yet available
panic: proto: illegal wireType 6

goroutine 1160 [running]:
panic(0x1a868e0, 0xc420982bc0)
	/usr/local/go/src/runtime/panic.go:500 +0x1ae
github.com/coreos/etcd/raft.newRaft(0xc420277368, 0x0)
	/go/src/github.com/coreos/etcd/raft/raft.go:196 +0x15b9
github.com/coreos/etcd/raft.NewRawNode(0xc420277368, 0x0, 0x0, 0x0, 0xc420039340, 0xc420039348, 0xd4dd19)
	/go/src/github.com/coreos/etcd/raft/rawnode.go:76 +0x96
github.com/cockroachdb/cockroach/storage.(*Replica).withRaftGroupLocked(0xc420198000, 0xc420277838, 0xc420012558, 0x7f10c2b8c400)
	/go/src/github.com/cockroachdb/cockroach/storage/replica.go:332 +0x4b0
github.com/cockroachdb/cockroach/storage.(*Replica).handleRaftReady(0xc420198000, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/storage/replica.go:1553 +0x293
github.com/cockroachdb/cockroach/storage.(*Store).processRaft.func1.1(0xc4209618c0, 0xc420244080, 0xc420198000)
	/go/src/github.com/cockroachdb/cockroach/storage/store.go:2558 +0x8f
created by github.com/cockroachdb/cockroach/storage.(*Store).processRaft.func1
	/go/src/github.com/cockroachdb/cockroach/storage/store.go:2561 +0x40f

goroutine 1 [chan receive]:
testing.(*T).Run(0xc42007da40, 0x1befbf3, 0x17, 0x1e814b8, 0x95cb01)
	/usr/local/go/src/testing/testing.go:647 +0x56e
testing.RunTests.func1(0xc42007da40)
	/usr/local/go/src/testing/testing.go:793 +0xba
testing.tRunner(0xc42007da40, 0xc42004bb38)
	/usr/local/go/src/testing/testing.go:610 +0xca
testing.RunTests(0x1e82e78, 0x2723900, 0x12c, 0x12c, 0xc42014a888)
	/usr/local/go/src/testing/testing.go:799 +0x4bb
testing.(*M).Run(0xc42004bef0, 0xc420285ce0)
	/usr/local/go/src/testing/testing.go:743 +0x130
github.com/cockroachdb/cockroach/storage_test.TestMain(0xc42004bef0)
	/go/src/github.com/cockroachdb/cockroach/storage/main_test.go:57 +0x287
main.main()
	github.com/cockroachdb/cockroach/storage/_test/_testmain.go:662 +0x1b6

goroutine 17 [syscall, locked to thread]:
runtime.goexit()
	/usr/local/go/src/runtime/asm_amd64.s:2086 +0x1

goroutine 4 [chan receive]:
github.com/cockroachdb/cockroach/util/log.(*loggingT).flushDaemon(0x2a5df40)
	/go/src/github.com/cockroachdb/cockroach/util/log/clog.go:1015 +0x85
created by github.com/cockroachdb/cockroach/util/log.init.1
	/go/src/github.com/cockroachdb/cockroach/util/log/clog.go:598 +0xc3

goroutine 1154 [syscall, locked to thread]:
github.com/cockroachdb/cockroach/storage/engine._Cfunc_DBPut(0x7f10bfa740e0, 0xc4209827f0, 0x1, 0x1, 0x0, 0xc420982ad0, 0xa, 0x0, 0x0)
	??:0 +0x7b
github.com/cockroachdb/cockroach/storage/engine.dbPut(0x7f10bfa740e0, 0xc4209827f0, 0x1, 0x8, 0x1, 0x0, 0xc420982ad0, 0xa, 0xa, 0x78, ...)
	/go/src/github.com/cockroachdb/cockroach/storage/engine/rocksdb.go:1339 +0x285
github.com/cockroachdb/cockroach/storage/engine.(*RocksDB).Put(0xc420961020, 0xc4209827f0, 0x1, 0x8, 0x1, 0x0, 0xc420982ad0, 0xa, 0xa, 0xc4202d2480, ...)
	/go/src/github.com/cockroachdb/cockroach/storage/engine/rocksdb.go:429 +0xa5
github.com/cockroachdb/cockroach/storage/engine.InMem.Put(0xc420961020, 0xc4209827f0, 0x1, 0x8, 0x1, 0x0, 0xc420982ad0, 0xa, 0xa, 0x0, ...)
	<autogenerated>:91 +0xa7
github.com/cockroachdb/cockroach/storage/engine.mvccPutInternal(0x7f10c2b8c400, 0xc420012550, 0x7f10c2b8c468, 0xc420961020, 0x27462e0, 0xc420a532c0, 0x0, 0xc4209827f0, 0x1, 0x8, ...)
	/go/src/github.com/cockroachdb/cockroach/storage/engine/mvcc.go:1091 +0x1741
--
testing.tRunner(0xc420a90180, 0x1e814b8)
	/usr/local/go/src/testing/testing.go:610 +0xca
created by testing.(*T).Run
	/usr/local/go/src/testing/testing.go:646 +0x530

goroutine 1121 [select]:
github.com/cockroachdb/cockroach/storage.(*Store).Start.func6()
	/go/src/github.com/cockroachdb/cockroach/storage/store.go:1175 +0x1e1
github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker.func1(0xc420212000, 0xc420244020)
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:187 +0x8b
created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:188 +0x74

goroutine 1157 [select]:
github.com/cockroachdb/cockroach/storage.(*bookie).start.func1()
	/go/src/github.com/cockroachdb/cockroach/storage/reservation.go:275 +0x4c8
github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker.func1(0xc420212000, 0xc420992da0)
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:187 +0x8b
created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:188 +0x74

goroutine 1155 [chan receive]:
github.com/cockroachdb/cockroach/rpc.NewContext.func1()
	/go/src/github.com/cockroachdb/cockroach/rpc/context.go:115 +0x85
github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker.func1(0xc420212000, 0xc4209921e0)
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:187 +0x8b
created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:188 +0x74

goroutine 1119 [select]:
github.com/cockroachdb/cockroach/storage.(*Store).startGossip.func2()
	/go/src/github.com/cockroachdb/cockroach/storage/store.go:1238 +0x2e9
github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker.func1(0xc420212000, 0xc4202f0060)
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:187 +0x8b
created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:188 +0x74

goroutine 1156 [chan receive]:
github.com/cockroachdb/cockroach/storage/engine.(*RocksDB).Open.func1(0xc420961020)
	/go/src/github.com/cockroachdb/cockroach/storage/engine/rocksdb.go:384 +0x64
created by github.com/cockroachdb/cockroach/storage/engine.(*RocksDB).Open
	/go/src/github.com/cockroachdb/cockroach/storage/engine/rocksdb.go:385 +0xb61

goroutine 1118 [select]:
github.com/cockroachdb/cockroach/storage.(*Store).startGossip.func1()
	/go/src/github.com/cockroachdb/cockroach/storage/store.go:1219 +0x2e9
github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker.func1(0xc420212000, 0xc4202f0040)
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:187 +0x8b
created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:188 +0x74

goroutine 1120 [select]:
github.com/cockroachdb/cockroach/storage.(*Store).Start.func5()
	/go/src/github.com/cockroachdb/cockroach/storage/store.go:1165 +0x1e1
github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker.func1(0xc420212000, 0xc420244000)
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:187 +0x8b
created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:188 +0x74

goroutine 1161 [runnable]:
github.com/cockroachdb/cockroach/storage.(*Store).processRaft.func1.1(0xc4209618c0, 0xc420244080, 0xc420198480)
	/go/src/github.com/cockroachdb/cockroach/storage/store.go:2555
created by github.com/cockroachdb/cockroach/storage.(*Store).processRaft.func1
	/go/src/github.com/cockroachdb/cockroach/storage/store.go:2561 +0x40f

goroutine 1117 [select]:
github.com/cockroachdb/cockroach/storage.(*Store).Start.func4()
	/go/src/github.com/cockroachdb/cockroach/storage/store.go:1145 +0x19d
github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker.func1(0xc420212000, 0xc4202f0020)
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:187 +0x8b
created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:188 +0x74

goroutine 1162 [syscall, locked to thread]:
github.com/cockroachdb/cockroach/storage/engine._Cfunc_DBApplyBatchRepr(0x7f10bfa740e0, 0xc420215c00, 0x7e, 0x0, 0x0)
	??:0 +0x75
github.com/cockroachdb/cockroach/storage/engine.dbApplyBatchRepr(0x7f10bfa740e0, 0xc420215c00, 0x7e, 0x400, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/storage/engine/rocksdb.go:1354 +0xf2
github.com/cockroachdb/cockroach/storage/engine.(*RocksDB).ApplyBatchRepr(0xc420961020, 0xc420215c00, 0x7e, 0x400, 0x0, 0xc4200f35e8)
	/go/src/github.com/cockroachdb/cockroach/storage/engine/rocksdb.go:448 +0x6f
github.com/cockroachdb/cockroach/storage/engine.(*rocksDBBatch).Commit(0xc4201ba1c0, 0xc420012558, 0x2743cc0)
	/go/src/github.com/cockroachdb/cockroach/storage/engine/rocksdb.go:992 +0x52b
github.com/cockroachdb/cockroach/storage.(*Replica).handleRaftReady(0xc420090240, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/storage/replica.go:1593 +0x5a7
github.com/cockroachdb/cockroach/storage.(*Store).processRaft.func1.1(0xc4209618c0, 0xc420244080, 0xc420090240)
	/go/src/github.com/cockroachdb/cockroach/storage/store.go:2558 +0x8f
created by github.com/cockroachdb/cockroach/storage.(*Store).processRaft.func1
	/go/src/github.com/cockroachdb/cockroach/storage/store.go:2561 +0x40f

goroutine 1158 [semacquire]:
sync.runtime_Semacquire(0xc42024408c)
	/usr/local/go/src/runtime/sema.go:47 +0x30
sync.(*WaitGroup).Wait(0xc420244080)
	/usr/local/go/src/sync/waitgroup.go:131 +0xbf
github.com/cockroachdb/cockroach/storage.(*Store).processRaft.func1()
	/go/src/github.com/cockroachdb/cockroach/storage/store.go:2563 +0x43d
github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker.func1(0xc420212000, 0xc420992e00)
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:187 +0x8b
created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:188 +0x74
FAIL	github.com/cockroachdb/cockroach/storage	2.334s
=== RUN   TestBatchBasics
I160808 20:17:33.553016 storage/engine/rocksdb.go:353  opening in memory rocksdb instance
I160808 20:17:33.555615 storage/engine/rocksdb.go:398  closing in-memory rocksdb instance
--- PASS: TestBatchBasics (0.03s)
=== RUN   TestBatchRepr
I160808 20:17:33.586232 storage/engine/rocksdb.go:353  opening in memory rocksdb instance
I160808 20:17:33.589340 storage/engine/rocksdb.go:398  closing in-memory rocksdb instance
--- PASS: TestBatchRepr (0.02s)
=== RUN   TestBatchGet
I160808 20:17:33.603424 storage/engine/rocksdb.go:353  opening in memory rocksdb instance

```
Please assign, take a look and update the issue accordingly.

Added some additional logging and we're the RaftHardState and almost immediately reading it and what we read differs from what we wrote:

```
I160808 17:34:01.355669 storage/engine/mvcc.go:480  mvcc put /Local/RangeID/1/u/RaftHardState: 24 cf f6 f7 03 08 05 10 00 18 0a
panic: /Local/RangeID/1/u/RaftHardState: 00 00 00 00 03 76 61 6c 75 65: proto: illegal wireType 6
```

The hex is the encoded proto. Notice that the value we read (contained in the panic) doesn't even appear to have an initialized checksum. Something fairly fundamental is going wrong here.

--- a/storage/replica_data_iter_test.go
+++ b/storage/replica_data_iter_test.go
@@ -137,10 +137,14 @@ func TestReplicaDataIteratorEmptyRange(t *testing.T) {
 // it verifies the pre and post ranges still contain the expected data.
 func TestReplicaDataIterator(t *testing.T) {
 	defer leaktest.AfterTest(t)()
+	ctx := TestStoreContext()
+	// Disable Raft processing for this test as it mucks with low-level details
+	// of replica storage in an unsafe way.
+	ctx.TestingKnobs.DisableProcessRaft = true
 	tc := testContext{
 		bootstrapMode: bootstrapRangeOnly,
 	}
-	tc.Start(t)
+	tc.StartWithStoreContext(t, ctx)
 	defer tc.Stop()
 
 	// See notes in EmptyRange test method for adjustment to descriptor.
@@ -196,7 +200,7 @@ func TestReplicaDataIterator(t *testing.T) {
 			t.Fatal(err)
 		}
 		if bytes.HasPrefix(iter.Key().Key, unreplicatedPrefix) {
-
+			t.Fatalf("unexpected unreplicated key: %s", iter.Key().Key)
 		}
 	}
 

--- a/storage/store.go
+++ b/storage/store.go
@@ -565,6 +565,8 @@ type StoreTestingKnobs struct {
 	// DisableRefreshReasonTicks disables refreshing pending commands
 	// periodically.
 	DisableRefreshReasonTicks bool
+	// DisableProcessRaft disables the process raft loop.
+	DisableProcessRaft bool
 }
 
 var _ base.ModuleTestingKnobs = &StoreTestingKnobs{}
@@ -2511,6 +2513,9 @@ func (s *Store) enqueueRaftUpdateCheck(rangeID roachpb.RangeID) {
 // appropriate range. This method starts a goroutine to process Raft
 // commands indefinitely or until the stopper signals.
 func (s *Store) processRaft() {
+	if s.ctx.TestingKnobs.DisableProcessRaft {
+		return
+	}
 	sem := makeSemaphore(storeReplicaRaftReadyConcurrency)
 
 	s.stopper.RunWorker(func() {

